File: C/Insertion_sort.c
// بسم الله الرحمن الرحيم
// "وَأَنْ لَيْسَ لِلْإِنسَانِ إِلَّا مَا سَعَىٰ"
#include <stdio.h>

// Time Complexity O(n ^ 2)
void Insertion_sort(int nums[], int n)
{
    for (int i = 1; i < n; i++)
    {
        int val = nums[i];
        int j = i - 1;
        while (j >= 0 && nums[j] > val)
        {
            nums[j + 1] = nums[j];
            j--;
        }
        nums[j + 1] = val;
    }
}

int main()
{
    int n;
    scanf("%d", &n);
    int nums[n];
    for (int i = 0; i < n; i++)
        scanf("%d", &nums[i]);

    Insertion_sort(nums, n);

    for (int i = 0; i < n; i++)
    {
        printf("%d ", nums[i]);
    }
    printf("\n");

    return 0;
}

File: Python/Insertion_sort.py
# بسم الله الرحمن الرحيم
# "وَأَنْ لَيْسَ لِلْإِنسَانِ إِلَّا مَا سَعَىٰ"


# Time complexity: O(n^2)
def Insertion_sort(nums):
    n = len(nums)
    for i in range(1, n):
        val = nums[i]
        j = i - 1
        while j >= 0 and val < nums[j]:
            nums[j + 1] = nums[j]
            j -= 1
        nums[j + 1] = val


def main():
    n = int(input())
    nums = list(map(int, input().split()))
    Insertion_sort(nums)
    print(*nums)


if __name__ == "__main__":
    main()
